File: src/divide/dividing.cpp
/** Functions to speed up dividing and splitting algorithms in Estnltk's Text class

These are meant to be used for splitting layers.
Has optimized code for both simple and multi layers.
Duplicates the functionality of pure-Python divide module.
*/
#include "dividing.h"

// define values representing None
const Span none(-1,-1);
const SpanVector<Span>(1, none) vec_none;


/// return the first element of a simple span
int first(Span const& span) {
    return span.first;
}

/// return the first element of a multi span
int first(SpanVector const& spans) {
    return spans[0].first;
}

/// last element of a simple span
int last(Span const& span) {
    return span.second;
}

/// last element of a multi span
int last(SpanVector const& spans) {
    return spans[spans.size()-1].second;
}

/// Remove duplicate integers.
/// Resulting list might not be sorted.
IntVector unique(IntVector& const vec) {
    std::unordered_set<int> seen;
    IntVector unique_vec;
    unique_vec.reserve(vec.size());
    for (int i = 0 ; i < vec.size() ; ++i) {
        int e = vec[i];
        if (seen.find(e) == seen.end()) {
            seen.insert(e);
            unique_vec.push_back(e);
        }
    }
    return unique_vec;
}


/// True if the simple outer span contains the simple inner span.
bool span_contains_span(Span const& outer, Span const& inner) {
    return outer.first <= inner.first and outer.second >= inner.second;
}


/// True if simple outline contains the multi inner span.
bool span_contains_list(Span const& outer, SpanVector const& inner) {
    for (int i = 0 ; i < inner.size() ; ++i) {
        if (!span_contains_span(outer, inner[i])) {
            return false;
        }
    }
    return true;
}

/// True of multi outline contains simple inner span.
bool list_contains_span(SpanVector const& outer, Span const& inner) {
    for (int i = 0 ; i < outer.size() ; ++i) {
        if (span_contains_span(outer[i], inner)) {
            return true;
        }
    }
    return false;
}

/// True of multi outer span contains multi inner span.
bool list_contains_list(SpanVector const& outer, SpanVector const& inner) {
    const int n = outer.size();
    const int m = inner.size();
    int i = 0;
    int j = 0;
    while (i < n && j < m) {
        if (outer[i].first > inner[j].first) {
            return false;
        }
        if (outer[i].second >= inner[j].second) {
            j += 1;
        } else {
            i += 1;
        }
    }
    return j == m;
}

/// Create a simple span with positions relative to outer span start.
Span span_translates_span(Span const& outer, Span const& inner) {
    if (span_contains_span(outer, inner)) {
        Span span(inner.first - outer.first, inner.second - outer.first);
    }
    return none;
}

/// Create a multi span with positions relative to outer span start.
SpanVector span_translates_list(Span const& outer, SpanVector const& inner) {
    const int left = outer.first;
    SpanVector translated;
    translated.reserve(inner.size());
    for (int i = 0 ; i < inner.size() ; ++i) {
        if (span_contains_span(outer, inner[i])) {
            Span span(inner[i].first - left, inner[i].second - left);
            translated.push_back(span);
        }
    }
    return translated;
}

/// Create a simple span with positions relative to outer span start.
Span list_translates_span(SpanVector const& outer, Span const& inner, std::string const& sep) {
    const int seplen = sep.size();
    int offset = 0;
    for (int i = 0 ; i < outer.size() ; ++i) {
        if (span_contains_span(outer[i], inner)) {
            Span span(inner.first - outer[i].first + offset, inner.second - outer[i].first + offset);
            return span;
        }
        offset += outer[i].second - outer[i].first + seplen;
    }
}

/// Create a multi span with positions relative to outer span start.
SpanVector list_translates_list(SpanVector const& outer, SpanVector const& inner, std::string const& sep) {
    const int n = outer.size();
    const int m = inner.size();
    const int seplen = sep.size();
    int i = 0;
    int j = 0;
    int offset = 0;
    SpanVector translated;
    translated.reserve(n + m);
    while (i < n && j < m) {
        Span const& ospan = outer[i];
        Span const& ispan = inner[j];
        if (ospan.first > ispan.first) {
            j += 1;
            continue;
        }
        if (ospan.second >= ispan.second) {
            Span span(ispan.first - ospan.first + offset, ispan.second - ospan.first + offset);
            j += 1;
        } else {
            offset += ospan.second - ospan.first + seplen;
        }
    }
    if (translated.size() > 0) {
        return translated;
    }
    return vec_none;
}

/// Compute the mapping of which outer spans contain which inner spans.
IntVectors spans_collect_spans(SpanVector const& outer_spans, SpanVector const& inner_spans) {
    const int n = outer_spans.size();
    const int m = inner_spans.size();
    int i = 0;
    int j = 0;
    IntVectors intvectors;
    intvectors.reserve(n);
    IntVector current_bin;
    while (i < n && j < m) {
        Span const& ospan = outer_spans[i];
        Span const& ispan = inner_spans[j];
        if (ospan.first > ispan.first) {
            j += 1;
            continue;
        }
        if (ospan.second >= ispan.second) {
            current_bin.append(j);
            j += 1;
        } else {
            intvectors.append(current_bin);
            current_bin.clear();
            i += 1;
        }
    }
    if (intvectors.size() < n) {
        intvectors.append(current_bin);
        current_bin.clear();
    }
    while (intvectors.size() < n) {
        intvectors.append(current_bin);
    }
    return intvectors;
}


/// Compute the mapping of which outer spans contain which inner spans.
IntVectors spans_collect_lists(SpanVector const& outer_spans, SpanVectors const& inner_spans) {
    const int n = inner_spans.size();
    IndexedSpanVector flattened_spans;
    flattened_spans.reserve(n);

    for (int i = 0 ; i < n ; ++i) {
        SpanVector& spans = inner_spans[i];
        const int m = spans.size();
        for (int j = 0; j < m ; ++j) {
            flattened_spans.push_back(IndexedSpan(spans[j], i));
        }
    }
    sort(flattened_spans.first(), flattened_spans.last());

    IntVectors bins = spans_collect_spans(outer_spans, flattened_spans);
    for (int i = 0 ; i < bins.size() ; ++i) {
        IntVector& vec = bins[i];
        for (j = 0 ; j < vec.size() ; ++ j) {
            vec[j] = flattened_spans[vec[j]].second;
        }
        bins[i] = unique(vec);
    }
    return bins;
}

/// Compute the mapping of which outer spans contain which inner spans.
IntVectors lists_collect_spans(SpanVectors const& outer_spans, SpanVector const& inner_spans) {
    const int n = outer_spans.size();
    IndexedSpanVector flattened_spans;
    flattened_spans.reserve(n);

    for (int i = 0 ; i < n ; ++i) {
        SpanVector& spans = outer_spans[i];
        const int m = spans.size();
        for (int j = 0; j < m ; ++j) {
            flattened_spans.push_back(IndexedSpan(spans[j], i));
        }
    }
    sort(flattened_spans.first(), flattened_spans.last());

    IntVectors flat_bins = spans_collect_spans(flattened_spans, inner_spans);
    IntVectors bins(n);
    for (int i = 0 ; i < bins.size() ; ++i) {
        const int binidx = flattened_spans[i].second;
        IntVector& bin = bins[i];
        IntVector& flat_bin = flat_bins[j];
        bin.insert(bin.end(), flat_bin.begin(), flat_bin.end());
    }
    for (int i = 0 ; i < bins.size() ; ++i) {
        bins[i] = unique(bins[i]);
    }
    return bins;
}

/// Compute the mapping of which outer spans contain which inner spans.
IntVectors lists_collect_lists(SpanVectors const& outer_spans, SpanVector const& inner_spans) {
    const int n = outer_spans.size();
    IndexedSpanVector flattened_spans;
    flattened_spans.reserve(n);

    for (int i = 0 ; i < n ; ++i) {
        SpanVector& spans = outer_spans[i];
        const int m = spans.size();
        for (int j = 0; j < m ; ++j) {
            flattened_spans.push_back(IndexedSpan(spans[j], i));
        }
    }
    sort(flattened_spans.first(), flattened_spans.last());

    IntVectors flat_bins = spans_collect_lists(flattened_spans, inner_spans);
    IntVectors bins(n);
    for (int i = 0 ; i < bins.size() ; ++i) {
        const int binidx = flattened_spans[i].second;
        IntVector& bin = bins[i];
        IntVector& flat_bin = flat_bins[j];
        bin.insert(bin.end(), flat_bin.begin(), flat_bin.end());
    }
    for (int i = 0 ; i < bins.size() ; ++i) {
        bins[i] = unique(bins[i]);
    }
    return bins;
}

File: estnltk/legacy/dividing.py
# -*- coding: utf-8 -*-
"""
Module containing functions for splitting and dividing spans.
These are mainly used in :py:class:`~estnltk.text.Text` class split and divide methods.
"""
from __future__ import unicode_literals, print_function, absolute_import
from .names import START, END
from copy import deepcopy


def span_contains_span(outer, inner):
    return outer[0] <= inner[0] and outer[1] >= inner[1]


def span_contains_list(outer, inner):
    return all(span_contains_span(outer, span) for span in inner)


def list_contains_span(outer, inner):
    return any(span_contains_span(span, inner) for span in outer)


def list_contains_list(outer, inner):
    n, m = len(outer), len(inner)
    i, j = 0, 0
    while i < n and j < m:
        (ostart, oend), (istart, iend) = outer[i], inner[j]
        if ostart > istart:
            return False
        if oend >= iend:
            j += 1
        else:
            i += 1
    return j == m


def contains(outer, inner):
    inner_is_list = isinstance(inner, list)
    outer_is_list = isinstance(outer, list)
    if inner_is_list:
        if outer_is_list:
            return list_contains_list(outer, inner)
        return span_contains_list(outer, inner)
    if outer_is_list:
        return list_contains_span(outer, inner)
    return span_contains_span(outer, inner)


def any_filters_span(outer, inner):
    if contains(outer, inner):
        return inner


def span_filters_list(outer, inner):
    filtered = [span for span in inner if span_contains_span(outer, span)]
    if len(filtered) > 0:
        return filtered


def list_filters_list(outer, inner):
    n, m = len(outer), len(inner)
    i, j = 0, 0
    filtered = []
    while i < n and j < m:
        (ostart, oend), (istart, iend) = outer[i], inner[j]
        if ostart > istart:
            j += 1
            continue
        if oend >= iend:
            filtered.append((istart, iend))
            j += 1
        else:
            i += 1
    if len(filtered) > 0:
        return filtered


def span_translates_span(outer, inner):
    if span_contains_span(outer, inner):
        outer_left = outer[0]
        return inner[0]-outer_left, inner[1]-outer_left


def span_translates_list(outer, inner):
    outer_left = outer[0]
    translated = [(span[0]-outer_left, span[1]-outer_left) for span in inner if span_contains_span(outer, span)]
    if len(translated) > 0:
        return translated


def list_translates_span(outer, inner, sep):
    offset = 0
    seplen = len(sep)
    for span in outer:
        if span_contains_span(span, inner):
            outer_left = first(span)
            return inner[0]-outer_left+offset, inner[1]-outer_left+offset
        offset += span[1]-span[0] + seplen


def list_translates_list(outer, inner, sep):
    n, m = len(outer), len(inner)
    i, j = 0, 0
    translated = []
    offset = 0
    seplen = len(sep)
    while i < n and j < m:
        (ostart, oend), (istart, iend) = outer[i], inner[j]
        if ostart > istart:
            j += 1
            continue
        if oend >= iend:
            outer_left = first(outer[i])
            translated.append((istart-outer_left+offset, iend-outer_left+offset))
            j += 1
        else:
            offset += oend - ostart + seplen
            i += 1
    if len(translated) > 0:
        return translated


def filter_containing(outer, inner, translate=False, sep=' '):
    inner_is_list = isinstance(inner, list)
    outer_is_list = isinstance(outer, list)

    if translate:
        if inner_is_list:
            if outer_is_list:
                return list_translates_list(outer, inner, sep)
            return span_translates_list(outer, inner)
        if outer_is_list:
            return list_translates_span(outer, inner, sep)
        return span_translates_span(outer, inner)
    if inner_is_list:
        if outer_is_list:
            return list_filters_list(outer, inner)
        return span_filters_list(outer, inner)
    return any_filters_span(outer, inner)


def spans(element):
    start = element[START]
    end = element[END]
    return convert_span((start, end))


def convert_span(span):
    # should be already in correct format
    if isinstance(span, list):
        return span
    # otherwise detect, if it is a simple span or multispan
    start, end = span
    if isinstance(start, int):
        return int(start), int(end)
    return list(zip(start, end))


def first(span):
    if isinstance(span, list):
        return span[0][0]
    return span[0]


def last(span):
    if isinstance(span, list):
        return span[-1][1]
    return span[1]


def update_span(element, spans):
    if isinstance(spans, list):
        starts, ends = zip(*spans)
        element[START] = list(starts)
        element[END] = list(ends)
    else:
        element[START] = spans[0]
        element[END] = spans[1]
    return element


def spans_collect_spans(outer_spans, inner_spans):
    n, m = len(outer_spans), len(inner_spans)
    i, j = 0, 0
    current_bin = []
    nyielded = 0
    while i < n and j < m:
        (ostart, oend), (istart, iend) = outer_spans[i], inner_spans[j]
        if ostart > istart:
            j += 1
            continue
        if oend >= iend:
            current_bin.append(j)
            j += 1
        else:
            yield current_bin
            nyielded += 1
            current_bin = []
            i += 1
    # yield the last (possibly) half-filled bin
    if nyielded < n:
        yield current_bin
        nyielded += 1
    # yield empty bins
    while nyielded < n:
        yield []
        nyielded += 1


def unique(iterable):
    seen = set()
    for e in iterable:
        if e not in seen:
            yield e
            seen.add(e)


def spans_collect_lists(outer_spans, inner_spans):
    mapping = []
    flattened_spans = []
    for idx, spans in enumerate(inner_spans):
        for s in spans:
            flattened_spans.append(s)
            mapping.append(idx)
    flattened_spans, mapping = zip(*sorted(zip(flattened_spans, mapping)))
    for bin in spans_collect_spans(outer_spans, flattened_spans):
        yield unique(mapping[idx] for idx in bin)


def lists_collect_spans(outer_spans, inner_spans):
    mapping = []
    flattened_spans = []
    for idx, spans in enumerate(outer_spans):
        for s in spans:
            flattened_spans.append(s)
            mapping.append(idx)
    flattened_spans, mapping = zip(*sorted(zip(flattened_spans, mapping)))
    flat_bins = list(spans_collect_spans(flattened_spans, inner_spans))
    bins = [[] for _ in range(len(outer_spans))]
    for flatidx, flatbin in enumerate(flat_bins):
        binidx = mapping[flatidx]
        bins[binidx].extend(flatbin)
    for bin in bins:
        yield unique(bin)


def lists_collect_lists(outer_spans, inner_spans):
    mapping = []
    flattened_spans = []
    for idx, spans in enumerate(outer_spans):
        for s in spans:
            flattened_spans.append(s)
            mapping.append(idx)
    flattened_spans, mapping = zip(*sorted(zip(flattened_spans, mapping)))
    flat_bins = list(spans_collect_lists(flattened_spans, inner_spans))
    bins = [[] for _ in range(len(outer_spans))]
    for flatidx, flatbin in enumerate(flat_bins):
        binidx = mapping[flatidx]
        bins[binidx].extend(flatbin)
    for bin in bins:
        yield unique(bin)


def get_bins(outer_spans, inner_spans):
    outers_are_lists = isinstance(outer_spans[0], list)
    inners_are_lists = isinstance(inner_spans[0], list)
    if outers_are_lists:
        if inners_are_lists:
            return lists_collect_lists(outer_spans, inner_spans)
        return lists_collect_spans(outer_spans, inner_spans)
    else:
        if inners_are_lists:
            return spans_collect_lists(outer_spans, inner_spans)
        return spans_collect_spans(outer_spans, inner_spans)


def get_filterer(outer_spans, inner_spans, translate, sep):
    outers_are_lists = isinstance(outer_spans[0], list)
    inners_are_lists = isinstance(inner_spans[0], list)

    if translate:
        if inners_are_lists:
            if outers_are_lists:
                return lambda outer, inner: list_translates_list(outer, inner, sep)
            return lambda outer, inner: span_translates_list(outer, inner)
        if outers_are_lists:
            return lambda outer, inner: list_translates_span(outer, inner, sep)
        return lambda outer, inner: span_translates_span(outer, inner)
    if inners_are_lists:
        if outers_are_lists:
            return lambda outer, inner: list_filters_list(outer, inner)
        return lambda outer, inner: span_filters_list(outer, inner)
    return lambda outer, inner: any_filters_span(outer, inner)


def divide_by_spans(elements, outer_spans, translate=False, sep=' '):
    outer_spans = [convert_span(s) for s in outer_spans]
    inner_spans = [spans(e) for e in elements]
    if len(inner_spans) == 0:
        return [[] for _ in range(len(outer_spans))]
    if len(outer_spans) == 0:
        return []
    inners_are_lists = isinstance(inner_spans[0], list)
    filterer = get_filterer(outer_spans, inner_spans, translate, sep)
    bins = []
    for binidx, collection in enumerate(get_bins(outer_spans, inner_spans)):
        outer = outer_spans[binidx]
        bin = []
        for elemidx in collection:
            elem = elements[elemidx]
            filtered = filterer(outer, inner_spans[elemidx])
            if filtered is not None:
                if inners_are_lists or translate:
                    elem = deepcopy(elem)
                elem = update_span(elem, filtered)
                bin.append(elem)
        bins.append(bin)
    return bins


def divide_by_spans_old_inefficient(elements, outer_spans, translate=False, sep=' '):
    inner_elems = elements
    outer_spans = [(i, convert_span(e)) for i, e in enumerate(outer_spans)]
    inner_spans = [(i, spans(e)) for i, e in enumerate(inner_elems)]
    bins = [[] for _ in range(len(outer_spans))]
    for iidx, ispan in inner_spans:
        for oidx, ospan in outer_spans:
            filtered = filter_containing(ospan, ispan, translate, sep)
            if filtered is not None:
                elem = inner_elems[iidx]
                if isinstance(ispan, list) or translate:
                    elem = deepcopy(elem)
                elem = update_span(elem, filtered)
                bins[oidx].append(elem)
    return bins


def divide(elements, by, translate=False, sep=' '):
    """Divide lists `elements` and `by`.
    All elements are grouped into N bins, where N denotes the elements in `by` list.

    Parameters
    ----------
    elements: list of dict
        Elements to be grouped into bins.
    by: list of dict
        Elements defining the bins.
    translate: bool (default: False)
        When dividing, also translate start and end positions of elements.
    sep: str (default ' ')
        In case of multispans, what is the default text separator.
        This is required in order to tag correct start, end positions of elements.
    """
    outer_spans = [spans(elem) for elem in by]
    return divide_by_spans(elements, outer_spans, translate=translate, sep=sep)
